---
title: "s02_explore_annotations_add_hwe"
output: html_document
editor_options: 
  chunk_output_type: console
---

started: AL10Dec2018  
last updated: AL30Sep2019  

# Summary

Explore variants annotations and add HWE data  

This changes number of annotations from 101 to 102  

Only 92 of 3,765 variants have high impact  

Somehow 45 of 3,765 variants happend to be down-sampled in at least one sample,  
despite my explicit request fo not down-sampling during the dataanalysis.  
However, this was a significant drop comparatively to Ampliseq-3 analysis,  
where I did not suppress down-sampling and had 4,009 out of 4,164 variants down-sampled.  

Because Inf was excluded in -log(ExcessHets) plotting, 
the highest ExcessHets was not in the plots for variants violating HWE.  

7 variants grossly violating HWE (p<10-5) have been excluded.  

Input data: 3,765 vars x 535 Samples  
Output data: 3,758 vars x 535 Samples  

# start section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r set_width, echo=F}

# eval=F

options(width = 999)
# https://stackoverflow.com/questions/36845178/width-of-r-code-chunk-output-in-rmarkdown-files-knitr-ed-to-html

```

```{r start_section}

Sys.time()
rm(list=ls())
graphics.off()

library(knitr)
library(dplyr)
library(tidyr) # for separate (parsing of cDNA, CDC and prot positions)

base_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s08_reshape_variants_annotations"
opts_knit$set(root.dir = base_folder)

options(stringsAsFactors = F)
options(warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

library(HardyWeinberg) # for calculating HWE  

#options(error = browser()) # Type Q or c to exit, drop browser level
# https://support.rstudio.com/hc/en-us/articles/200713843?version=1.1.456&mode=desktop
# https://stackoverflow.com/questions/13052522/how-to-leave-the-r-browser-mode-in-the-console-window/13052588 

```

# read data

```{r read_data}

load(paste(base_folder, "s01_clean_variants_annotations.RData", sep="/"))

```

# Check data

```{r start_check}

ls()
dim(gt.mx)
dim(vars.df)
sum(rownames(gt.mx) != rownames(vars.df))

```

# Accessory function to explore selected fields

```{r explore_udf}

explore_field.udf <- function(field, field_separator){
  a <- strsplit(as.vector(field),paste("\\",field_separator,sep=""))
  b <- as.vector(unlist(a))
  sort(table(b), decreasing = TRUE)
}

```

# List of fields

```{r var_annotations}

colnames(vars.df)
str(vars.df, list.len=ncol(vars.df))

```

# Explore various variant annotation fields  

```{r explore_vars}

# VEP consequence & Impact
explore_field.udf(vars.df$Consequence, "&")
sort(table(vars.df$IMPACT), decreasing=T)

# Were any of the variants downsampled?
table(vars.df$DS) # How some variants happened to be down-sampled ???

# InbreedingCoeff
hist(vars.df$InbreedingCoeff, ylim=c(0,3500), lab=T)

# ExcessHet
hist(vars.df$ExcessHet, ylim=c(0,3500), lab=T)

plot(-log10(vars.df$ExcessHet), main="- log10 ExcessHet", xlab="Variants")
abline(h=3.8, col="red", lty=2)
x <- which(-log10(vars.df$ExcessHet) > 3.8 & is.finite(-log10(vars.df$ExcessHet)))
y <- -log10(vars.df$ExcessHet)[x]
text(x, y, labels=vars.df$SplitVarID[x], 
     pos=c(2,2,1,4,2), cex=0.5, offset=0.3)
vars.df$SplitVarID[x]
# MLEAF
hist(as.numeric(vars.df$MLEAF),ylim=c(0,3600), lab=T)

# MQ
hist(vars.df$MQ,ylim=c(0,4000), lab=T)

# MultiAllelic
table(vars.df$Multiallelic)

# NEGATIVE_TRAIN_SITE
table(vars.df$NEGATIVE_TRAIN_SITE)

# POSITIVE_TRAIN_SITE
table(vars.df$POSITIVE_TRAIN_SITE)

# SIFT Polyphen
sort(table(vars.df$SIFT_call), decreasing=T)
sort(table(vars.df$PolyPhen_call), decreasing=T)

# VEP CLIN_SIG
sort(table(vars.df$CLIN_SIG), decreasing=T)[1:15]
sum(is.na(vars.df$CLIN_SIG))
explore_field.udf(vars.df$CLIN_SIG, "&")

# SYMBOL
sum(is.na(vars.df$SYMBOL))

# DISTANCE (Shortest distance from variant to transcript)
hist(as.numeric(vars.df$DISTANCE), lab=T, ylim=c(0,12))
table(as.numeric(vars.df$DISTANCE))

# STRAND
table(vars.df$STRAND)

# Clean-up
rm(x,y,explore_field.udf)

```

# Add HWE data to biallelic variants  

using library HardyWeinberg  

```{r hwe}

# Prepare genotypes counts
genotypes_counts <- MakeCounts(t(gt.mx),coding=c(0,1,2))
dim(genotypes_counts)
genotypes_counts[1:5,]

# Calculate HWE p-values
hwe <- HWExactStats(genotypes_counts[,1:3], verbose=FALSE)

# Remove HWE p from multiallelic variants
NA -> hwe[vars.df$Multiallelic]
sum(is.na(hwe))

# Plot HWE p-values
length(hwe)
names(hwe) <- rownames(gt.mx)
hwe[1:5] 
min(hwe, na.rm=T)
max(hwe, na.rm=T)
hist(hwe)

plot(-log10(hwe), main="- log10 HWE", xlab="Variants")
abline(h=5, col="red", lty=2)
x <- which(-log10(hwe) > 5)
y <- -log10(hwe)[x]
text(x, y, labels=names(hwe)[x], 
     pos=c(1,1,1,2,4,1,3), cex=0.5, offset=0.2)

# Add hwe to vars.df
vars.df <- cbind(vars.df, hwe_biallelic=hwe)
colnames(vars.df)

# Clean-up (keep hwe and vars_bi.df)
rm(genotypes_counts, hwe, x, y)

```

# Remove variants grossly violating HWE

```{r}

hwe_violators <- vars.df$hwe_biallelic < 1e-5 & !is.na(vars.df$hwe_biallelic)
sum(hwe_violators)

vars.df[hwe_violators,c("hwe_biallelic","ExcessHet")]

gt.mx <- gt.mx[!hwe_violators,]
vars.df <- vars.df[!hwe_violators,]

rm(hwe_violators)

```

# Check data

```{r final_check}

ls()
dim(gt.mx)
dim(vars.df)
sum(rownames(gt.mx) != rownames(vars.df))

```

# Save result

```{r save_result}

save.image(paste(base_folder, "s02_explore_annotations_add_hwe.RData", sep="/"))

```

# final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
