---
title: "s01_add_PCs"
output: html_document
editor_options: 
  chunk_output_type: console
---

started: AL26Apr2019  
last updated: AL01Oct2019  

# Summary  

Input sequencing data: 3,675 vars x 515 samples (258UBC + 257CBC)  
Input eigenvectors for 515 samples (calculated using 468 non-rare variants not in LD) 

Add eigenvectors to phenotypes, make PCA plots and  
Mark 33 PC outliers: outside of mean +/- 3*sd on the top 2 eigenvectors

# Start_section

<style>
pre {
  overflow-x: auto;
}
pre code {
  word-wrap: normal;
  white-space: pre;
}
</style>

```{r set_width, echo=F}

#eval=F

options(width = 999)
# https://stackoverflow.com/questions/36845178/width-of-r-code-chunk-output-in-rmarkdown-files-knitr-ed-to-html

```

```{r start_section}

Sys.time()
rm(list=ls())
graphics.off()

library(knitr)
library(dplyr)
library(ggplot2)
library(plotly)

base_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s13_ampliseq_only_PCA/s03_explore_PCA_plots_exclude_outliers"

opts_knit$set(root.dir = base_folder)

options(stringsAsFactors = F)
options(warnPartialMatchArgs = T, 
        warnPartialMatchAttr = T, 
        warnPartialMatchDollar = T)

#options(error = browser()) # Type Q or c to exit, drop browser level
# https://support.rstudio.com/hc/en-us/articles/200713843?version=1.1.456&mode=desktop
# https://stackoverflow.com/questions/13052522/how-to-leave-the-r-browser-mode-in-the-console-window/13052588 

```

# Read_data  

```{r read_data}

# Threshold for outlier detection
th <- 3 # mean +/- (th*sd)

# Sequencing data (for wecare phenotypes: case/control status)
source_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s11_remove_BRCA_PALB_carriers"
load(paste(source_folder, "s01_exclude_BRCA1_BCRA2_PALB2_carriers.RData", sep="/"))
base_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s13_ampliseq_only_PCA/s03_explore_PCA_plots_exclude_outliers"

# Eigenvectors & eigenvalues
source_folder="/Users/alexey/Documents/wecare/ampliseq/v05_ampliseq_only/s13_ampliseq_only_PCA/s02_calculate_ampliseq_only_PCs/data/s04_pca"

eigenvectors_file <- paste(source_folder, "ampliseq_only_468_515_100PCs.eigenvec", sep="/")
eigenvalues_file <- paste(source_folder, "ampliseq_only_468_515_100PCs.eigenval", sep="/")

eigenvectors.df <- read.table(eigenvectors_file, header=T, sep="\t",quote="")
eigenvalues.df <- read.table(eigenvalues_file, header=F, sep="\t",quote="")

# Clean-up
rm(source_folder, eigenvectors_file, eigenvalues_file)

```

# Check data

```{r start_check}

# List of objects
ls()

# Dimentions of objects
dim(eigenvectors.df)
dim(eigenvalues.df)
dim(genotypes.mx)
dim(phenotypes.df)
dim(variants.df)

# Counts of nfe/controls/cases
table(phenotypes.df$cc)

# Update eigenvectors
rownames(eigenvectors.df) <- eigenvectors.df$FID
eigenvectors.df <- eigenvectors.df[,-1]
"long_ids" -> colnames(eigenvectors.df)[1]
eigenvectors.df[1:5,1:5]

```

# Plot eigenvalues

```{r eigenvalues}

plot(eigenvalues.df$V1, type="b", 
     xlab="Eigenvector", ylab="Eigenvalue",
     ylim=c(0,20),main="Top 100 eigenvectors")

plot(eigenvalues.df$V1[1:20], type="b", 
     xlab="Eigenvector", ylab="Eigenvalue",
     ylim=c(0,20), main="Top 20 eigenvectors")

rm(eigenvalues.df)

```

# Prepare data for plots

```{r prepare_data}

# Add factor column with verbal lables for phenotypes to be used in plot
group <- factor(phenotypes.df$cc, levels = c(0,1), labels = c("UBC","CBC"))
phenotypes.df <- data.frame(phenotypes.df,group)
table(phenotypes.df$group)

# Add 5 top eigenvectors to phenotypes
phenotypes.df <- full_join(phenotypes.df,eigenvectors.df[,1:6], by="long_ids")
rownames(phenotypes.df) <- phenotypes.df$long_ids

# Prepare colour scale for ggplots
myColours <- c("UBC"="blue", "CBC"="red")
myColourScale <- scale_colour_manual(values=myColours)

# Clean-up
rm(group, eigenvectors.df, myColours)

```

# Calculate thresholds for outliers

```{r calculate_thresholds}

pc1_mean <- mean(phenotypes.df$PC1)
pc1_sd <- sd(phenotypes.df$PC1)
pc1_min <- pc1_mean - pc1_sd * th
pc1_max <- pc1_mean + pc1_sd * th

pc2_mean <- mean(phenotypes.df$PC2)
pc2_sd <- sd(phenotypes.df$PC2)
pc2_min <- pc2_mean - pc2_sd * th
pc2_max <- pc2_mean + pc2_sd * th

pc3_mean <- mean(phenotypes.df$PC3)
pc3_sd <- sd(phenotypes.df$PC3)
pc3_min <- pc3_mean - pc3_sd * th
pc3_max <- pc3_mean + pc3_sd * th

pc4_mean <- mean(phenotypes.df$PC4)
pc4_sd <- sd(phenotypes.df$PC4)
pc4_min <- pc4_mean - pc4_sd * th
pc4_max <- pc4_mean + pc4_sd * th

pc5_mean <- mean(phenotypes.df$PC5)
pc5_sd <- sd(phenotypes.df$PC5)
pc5_min <- pc5_mean - pc5_sd * th
pc5_max <- pc5_mean + pc5_sd * th

rm(pc1_mean, pc1_sd, 
   pc2_mean, pc2_sd, 
   pc3_mean, pc3_sd, 
   pc4_mean, pc4_sd, 
   pc5_mean, pc5_sd,
   th)

```

# Mark outliers

```{r mark_outliers}

# Detect outliers on each PC
pc1_outlier <- phenotypes.df$PC1 < pc1_min | phenotypes.df$PC1 > pc1_max
pc2_outlier <- phenotypes.df$PC2 < pc2_min | phenotypes.df$PC2 > pc2_max
pc3_outlier <- phenotypes.df$PC3 < pc3_min | phenotypes.df$PC3 > pc3_max
pc4_outlier <- phenotypes.df$PC4 < pc4_min | phenotypes.df$PC4 > pc4_max
pc5_outlier <- phenotypes.df$PC5 < pc5_min | phenotypes.df$PC5 > pc5_max

# Detect outliers on any PC
pc_outlier <- pc1_outlier | pc2_outlier

#pc_outlier <- pc1_outlier | pc2_outlier | pc3_outlier | pc4_outlier | pc5_outlier

# Check counts of outliers
sum(pc1_outlier)
sum(pc2_outlier)
sum(pc3_outlier)
sum(pc4_outlier)
sum(pc5_outlier)

sum(pc_outlier)

# Add outliers data to phenotypes dataframe
phenotypes.df <- data.frame(phenotypes.df, pc_outlier, pc1_outlier, pc2_outlier,
                            pc3_outlier, pc4_outlier, pc5_outlier)

phenotypes.df %>%
  filter(pc_outlier) %>% 
  select(Sample_num, pc_outlier, pc1_outlier, pc2_outlier,
         pc3_outlier, pc4_outlier, pc5_outlier) %>% 
  arrange(as.integer(Sample_num))

# Clean-up
rm(pc_outlier, pc1_outlier, pc2_outlier, pc3_outlier, pc4_outlier, pc5_outlier)

```

# PC1 vs PC2

```{r pc1_pc2}

# Prepare ggplot
g <- ggplot(phenotypes.df, aes(PC1,PC2)) + 
  geom_point(aes(col=group, text=Sample_num)) +
  labs(title="PC1 vs PC2", x="PC1", y="PC2") +
  theme(legend.title=element_blank()) + # To suppress the legend title
  myColourScale +                       # otherwise it would be "group"
  geom_vline(xintercept=c(pc1_min, pc1_max), linetype="dotted", size=0.2) +
  geom_hline(yintercept=c(pc2_min, pc2_max), linetype="dotted", size=0.2)

# Draw static ggplot
#g

# Draw dynamic ggplotly
ggplotly(g, tooltip="text") # By default the tooltip would also show coordinates 

# Clean-up
rm(g)

```

# PC2 vs PC3

```{r pc2_pc3}

# Prepare ggplot
g <- ggplot(phenotypes.df, aes(PC2,PC3)) + 
  geom_point(aes(col=group, text=Sample_num)) +
  labs(title="PC2 vs PC3", x="PC2", y="PC3") +
  theme(legend.title=element_blank()) + 
  myColourScale + 
  geom_vline(xintercept=c(pc2_min, pc2_max), linetype="dotted", size=0.2)
# geom_hline(yintercept=c(pc3_min, pc3_max), linetype="dotted", size=0.2)

# Draw static ggplot
#g

# Draw dynamic ggplotly
ggplotly(g, tooltip="text") # By default the tooltip would also show coordinates 

# Clean-up
rm(g)

```

# PC3 vs PC4

```{r pc3_pc4}

# Prepare ggplot
g <- ggplot(phenotypes.df, aes(PC3,PC4)) + 
  geom_point(aes(col=group, text=Sample_num)) +
  labs(title="PC3 vs PC4", x="PC3", y="PC4") +
  theme(legend.title=element_blank()) + 
  myColourScale 
#  geom_vline(xintercept=c(pc3_min, pc3_max), linetype="dotted", size=0.2) +
#  geom_hline(yintercept=c(pc4_min, pc4_max), linetype="dotted", size=0.2)

# Draw static ggplot
#g

# Draw dynamic ggplotly
ggplotly(g, tooltip="text") # By default the tooltip would also show coordinates 

# Clean-up
rm(g)

```

# PC4 vs PC5

```{r pc4_pc5}

# Prepare ggplot
g <- ggplot(phenotypes.df, aes(PC4,PC5)) + 
  geom_point(aes(col=group, text=Sample_num)) +
  labs(title="PC4 vs PC5", x="PC4", y="PC5") +
  theme(legend.title=element_blank()) + 
  myColourScale
#  geom_vline(xintercept=c(pc4_min, pc4_max), linetype="dotted", size=0.2) +
#  geom_hline(yintercept=c(pc5_min, pc5_max), linetype="dotted", size=0.2)

# Draw static ggplot
#g

# Draw dynamic ggplotly
ggplotly(g, tooltip="text") # By default the tooltip would also show coordinates 

# Clean-up
rm(g, myColourScale, 
   pc1_min, pc1_max, pc2_min, pc2_max, pc3_min, pc3_max, pc4_min, pc4_max, pc5_min, pc5_max)

```

# Save results

```{r save_results}

save.image(paste(base_folder, "s01_add_PCs.RData", sep="/"))

```

# Final_section

```{r final_section}

ls()
sessionInfo()
Sys.time()

```
